provider/core/src/any.rs

```diff
@@ -0,0 +1,386 @@
+    /// ```
+    pub fn from_static_ref<Y>(static_ref: &'static Y) -> Self
+    where
+        Y: for<'a> Yokeable<'a>,
+    {
+        AnyPayload {
+            inner: AnyPayloadInner::StructRef(static_ref),
+            // Note: This records the Yokeable type rather than the DataMarker type,
+            // but that is okay since this is only for debugging
+            type_name: core::any::type_name::<Y>(),
+        }
+    }
+
+    /// Creates an `AnyPayload` from a [`DataPayload`] stored in an [`Rc`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    /// use icu_provider::hello_world::*;
+    /// use std::borrow::Cow;
+    /// use std::rc::Rc;
+    ///
+    /// let payload: DataPayload<HelloWorldV1Marker> =
+    ///     DataPayload::from_owned(HelloWorldV1 {
+    ///         message: Cow::Borrowed("Custom Hello World")
+    ///     });
+    /// let rc_payload = Rc::from(payload);
+    ///
+    /// let any_payload = AnyPayload::from_rc_payload(rc_payload);
+    ///
+    /// let payload: DataPayload<HelloWorldV1Marker> = any_payload.downcast()
+    ///     .expect("TypeId matches");
+    /// assert_eq!("Custom Hello World", payload.get().message);
+    /// ```
+    pub fn from_rc_payload<M>(rc_payload: Rc<DataPayload<M>>) -> Self
+    where
+        M: DataMarker + 'static,
+    {
+        AnyPayload {
+            inner: AnyPayloadInner::PayloadRc(rc_payload),
+            type_name: core::any::type_name::<M>(),
+        }
+    }
+}
+
+impl<M> DataPayload<M>
+where
+    M: DataMarker + 'static,
+{
+    /// Wraps this DataPayload in an `Rc` and returns it as an `AnyPayload`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    /// use icu_provider::hello_world::*;
+    /// use std::borrow::Cow;
+    /// use std::rc::Rc;
+    ///
+    /// let payload: DataPayload<HelloWorldV1Marker> =
+    ///     DataPayload::from_owned(HelloWorldV1 {
+    ///         message: Cow::Borrowed("Custom Hello World")
+    ///     });
+    ///
+    /// let any_payload = payload.wrap_into_any_payload();
+    ///
+    /// let payload: DataPayload<HelloWorldV1Marker> = any_payload.downcast()
+    ///     .expect("TypeId matches");
+    /// assert_eq!("Custom Hello World", payload.get().message);
+    /// ```
+    pub fn wrap_into_any_payload(self) -> AnyPayload {
+        AnyPayload {
+            inner: AnyPayloadInner::PayloadRc(Rc::from(self)),
+            type_name: core::any::type_name::<M>(),
+        }
+    }
+}
+
+impl DataPayload<AnyMarker> {
+    /// Transforms a type-erased `DataPayload<AnyMarker>` into a concrete `DataPayload<M>`.
+    #[inline]
+    pub fn downcast<M>(self) -> Result<DataPayload<M>, DataError>
+    where
+        M: DataMarker + 'static,
+        for<'a> YokeTraitHack<<M::Yokeable as Yokeable<'a>>::Output>: Clone,
+        M::Yokeable: ZeroCopyFrom<'static, M::Yokeable>,
+    {
+        self.try_unwrap_owned()?.downcast()
```

Comment:
question: are we ever in danger from users manually constructing non-owned anypayloads?

(i know the current code has this too)

Reply:
I added a test to demonstrate how one could trigger this code path to run.